### PR TITLE
Xtrig arg validate

### LINKED ISSUE #5448: Cylc Validate failed to catch bad offset
### Description
An invalid isodatetime in the [[xtriggers]] section can get past cylc validate and cause a suite to fail.
I'm not sure if `cylc validate` should be expected to catch this typo, but I will raise it just in case:
```
    [[xtriggers]]
        L1_trigger = wall_clock(offset=PT755MH):PT30S  # <-- bad offset
```
I meant to type `PT7H55M`, but instead made a typo and wrote an offset of `755MH`

My suite kept failing to start (it just launched immediately into a "stopped" state) and I couldn't figure out why.
I only realised my error when I checked the "Workflow log file"

```
2023-04-03T13:37:42Z ERROR - could not convert string to float: '755M'
    Traceback (most recent call last):
      File "/home/h03/fcm/cylc-8.1.2-2/lib/python3.9/site-packages/cylc/flow/cycling/iso8601.py", line 881, in interval_parse
        return _interval_parse(interval_string)
      File "/home/h03/fcm/cylc-8.1.2-2/lib/python3.9/site-packages/cylc/flow/cycling/iso8601.py", line 902, in _interval_parse
        return WorkflowSpecifics.interval_parser.parse(interval_string)
      File "/home/h03/fcm/cylc-8.1.2-2/lib/python3.9/site-packages/metomi/isodatetime/parsers.py", line 575, in parse
        value = float(value)
    ValueError: could not convert string to float: '755M'
```

Cylc validate usually shouts about these kind of isodatetime errors, but not in this case.

### Reproducible Example

Add the above section into a flow.cylc and run as usual (I use `cylc vip`)
### Expected Behaviour

Suite will validate and install ok, but will immediately show in UI as "stopped".
Check workflow log, you should see the error above.

### PR DESCRIPTION
**Pull request by wxtim**
Close #5448 

Replaces https://github.com/cylc/cylc-flow/pull/5452, to which I have added tests.

From @hjoliver 
> This is not just a bug-fix for the `wall_clock` xtrigger - I've gone for a generic approach:
> - if an xtrigger module has a function called `validate` it will be called at validation time, with the parsed args.

From @MetRonnie (added to this PR)
> I realised it's trivially simple to automatically ensure xtrigger function args match the function signature at validation time

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/686.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by wxtim** [comment]
@markgrahamdawson  requesting a second review because this is quite collab between Hilary and I at this point.

**Review by MetRonnie — CHANGES_REQUESTED** [review]
I realised it's trivially simple to automatically ensure xtrigger function args match the function signature at validation time

PR open here: https://github.com/wxtim/cylc/pull/59

File: cylc/flow/xtriggers/wall_clock.py [hunk 1]
```diff
@@ -27,3 +29,44 @@ def wall_clock(trigger_time=None):
...
+        del f_args[0]
+
+    else:
+        # no args, infer zero interval
+        f_kwargs["offset"] = "P0Y"
+
+    # must be a valid interval
+    try:
+        interval_parse(f_kwargs["offset"])
+    except (ValueError, AttributeError):
```
[the remark this answers is not in the document]
Comment (wxtim) [reply cylc/flow/xtriggers/wall_clock.py:71]: I realized upon testing that we probably wanted to handle `f_kwargs["offset"]` not being a string. TBH, I'm not sure that it wouldn't be a case where `except Exception` would be justified - 99% of errors here are going to be user error, and for the 1% there's debug mode.

[the remark this answers is not in the document]
Comment (MetRonnie) [reply cylc/flow/xtriggers/wall_clock.py:71]: I'm not sure how `AttributeError` woild be raised, but you make a fair argument

**Review by MetRonnie — APPROVED** [review]

File: cylc/flow/xtriggers/xrandom.py [hunk 2]
```diff
@@ -89,3 +101,47 @@ def xrandom(percent, secs=0, _=None):
...
+def validate(f_args, f_kwargs, f_signature):
+    """Validate and manipulate args parsed from the workflow config.
+
+    percent: - 0 ≤ x ≤ 100
+    secs: An int.
+
+    If f_args used, convert to f_kwargs for clarity.
+
+    """
+    # convert to kwarg
```
Comment (MetRonnie) [inline cylc/flow/xtriggers/xrandom.py:115]:
```suggestion
    The rules for args are:
    * percent: Must be 0 ≤ x ≤ 100
    * secs: Must be an integer.
    """
    # convert to kwarg for clarity
```

**Comment by MetRonnie** [comment]
Flake8

**Comment by MetRonnie** [comment]
Woops, wrong button. Re-open when flake8 addressed

**Comment by wxtim** [comment]
> Woops, wrong button. Re-open when flake8 addressed

Fixed locally. Reopened so CI runs. Hopefully works on CI too.

**Review by MetRonnie — APPROVED** [review]

**Comment by MetRonnie** [comment]
I'm looking at the failing tests

**Review by MetRonnie — CHANGES_REQUESTED** [review]
While looking at the test failures, I realised the implementation of `xrandom`'s `validate` function was flawed with respect whether you used args vs kwargs.

Then I realised I could simplify the way validation is carried out, PR open at https://github.com/wxtim/cylc/pull/60

**Comment by MetRonnie** [comment]
Kicking tests

**Comment by MetRonnie** [comment]
Ah right, @wxtim when you squash merge a PR make sure you remove `[skip ci]` from the commit description otherwise the tests will not run

**Review by MetRonnie — APPROVED** [review]

**Comment by MetRonnie** [comment]
@wxtim Any objections if I merge this? I think enough due process has occurred at this point

**Comment by wxtim** [comment]
> Ah right, @wxtim when you squash merge a PR make sure you remove `[skip ci]` from the commit description otherwise the tests will not run

That seems reasonable - All the code has been reviewed by two pairs of eyes in effect even if the review was not linear.

**Pull request merged**